File: Module_3/49_concurrency_problem.cpp
// C++ Example - DATA RACE!
#include <thread>
#include <iostream>

int counter = 0;  // Shared mutable state

void increment() {
    for (int i = 0; i < 1000000; i++) {
        counter++;  // NOT ATOMIC!
    }
}

int main() {
    std::thread t1(increment);
    std::thread t2(increment);
    t1.join(); t2.join();
    
    // Result: ??? Not 2,000,000 (usually)
    std::cout << counter << std::endl;
    // Compiles fine, but UNDEFINED BEHAVIOR!
}

File: Module_3/49_concurrency_problem.rs
use std::thread;

// This WON'T COMPILE - ownership prevents data races
fn main() {
    let counter = 0;  // Shared state
    
    let handle = thread::spawn(move || {
        // counter moved here!
        // counter += 1;  // Wouldn't compile anyway
    });
    
    // println!("{}", counter);  // ERROR: counter moved!
}

// Rust's approach:
// 1. Ownership forces explicit sharing
// 2. Borrow checker prevents data races at compile time
// 3. Fearless concurrency: "If it compiles, it's race-free"
